src/content_mapblock.cpp

```diff
@@ -525,1633 +371,955 @@ static TileSpec getSpecialTile(const ContentFeatures &f,
+	tile_nocrack.material_flags &= ~MATERIAL_FLAG_CRACK;
+
+	// Put wood the right way around in the posts
+	TileSpec tile_rot = tile;
+	tile_rot.rotation = 1;
+
+	static const f32 post_rad = BS/8;
+	static const f32 bar_rad  = BS/16;
+	static const f32 bar_len  = BS/2 - post_rad;
+
+	// The post - always present
+	static const aabb3f post(-post_rad, -BS/2, -post_rad,
+	                          post_rad,  BS/2,  post_rad);
+	static const f32 postuv[24] = {
+		0.375, 0.375, 0.625, 0.625,
+		0.375, 0.375, 0.625, 0.625,
+		0.000, 0.000, 0.250, 1.000,
+		0.250, 0.000, 0.500, 1.000,
+		0.500, 0.000, 0.750, 1.000,
+		0.750, 0.000, 1.000, 1.000,
+	};
+	tile = tile_rot;
+	drawAutoLightedCuboid(post, postuv);
+
+	tile = tile_nocrack;
+
+	// Now a section of fence, +X, if there's a post there
+	v3s16 p2 = p;
+	p2.X++;
+	MapNode n2 = data->m_vmanip.getNodeNoEx(blockpos_nodes + p2);
+	const ContentFeatures *f2 = &nodedef->get(n2);
+	if (f2->drawtype == NDT_FENCELIKE) {
+		static const aabb3f bar_x1(BS/2 - bar_len,  BS/4 - bar_rad, -bar_rad,
+		                           BS/2 + bar_len,  BS/4 + bar_rad,  bar_rad);
+		static const aabb3f bar_x2(BS/2 - bar_len, -BS/4 - bar_rad, -bar_rad,
+		                           BS/2 + bar_len, -BS/4 + bar_rad,  bar_rad);
+		static const f32 xrailuv[24] = {
+			0.000, 0.125, 1.000, 0.250,
+			0.000, 0.250, 1.000, 0.375,
+			0.375, 0.375, 0.500, 0.500,
+			0.625, 0.625, 0.750, 0.750,
+			0.000, 0.500, 1.000, 0.625,
+			0.000, 0.875, 1.000, 1.000,
+		};
+		drawAutoLightedCuboid(bar_x1, xrailuv);
+		drawAutoLightedCuboid(bar_x2, xrailuv);
+	}
 
-			float d = (float)BS/64;
-			float s = BS/2;
+	// Now a section of fence, +Z, if there's a post there
+	p2 = p;
+	p2.Z++;
+	n2 = data->m_vmanip.getNodeNoEx(blockpos_nodes + p2);
+	f2 = &nodedef->get(n2);
+	if (f2->drawtype == NDT_FENCELIKE) {
+		static const aabb3f bar_z1(-bar_rad,  BS/4 - bar_rad, BS/2 - bar_len,
+		                            bar_rad,  BS/4 + bar_rad, BS/2 + bar_len);
+		static const aabb3f bar_z2(-bar_rad, -BS/4 - bar_rad, BS/2 - bar_len,
+		                            bar_rad, -BS/4 + bar_rad, BS/2 + bar_len);
+		static const f32 zrailuv[24] = {
+			0.1875, 0.0625, 0.3125, 0.3125, // cannot rotate; stretch
+			0.2500, 0.0625, 0.3750, 0.3125, // for wood texture instead
+			0.0000, 0.5625, 1.0000, 0.6875,
+			0.0000, 0.3750, 1.0000, 0.5000,
+			0.3750, 0.3750, 0.5000, 0.5000,
+			0.6250, 0.6250, 0.7500, 0.7500,
+		};
+		drawAutoLightedCuboid(bar_z1, zrailuv);
+		drawAutoLightedCuboid(bar_z2, zrailuv);
+	}
+}
 
-			short g = -1;
-			if (is_rail_x[4] || is_rail_x[5] || is_rail_z[4] || is_rail_z[5])
-				g = 1; //Object is at a slope
+bool MapblockMeshGenerator::isSameRail(v3s16 dir)
+{
+	MapNode node2 = data->m_vmanip.getNodeNoEx(blockpos_nodes + p + dir);
+	if (node2.getContent() == n.getContent())
+		return true;
+	const ContentFeatures &def2 = nodedef->get(node2);
+	return (def2.drawtype == NDT_RAILLIKE) &&
+		(def2.getGroup(raillike_groupname) 
```

Comment:
here it's a very good candidate to use function pointers in an array instead of doing a long switch case. please use a function pointer array, test the limits or invalid values, and call the pointed function instead this should improve the rendering performance